doc/monads/states.lean

```diff
@@ -0,0 +1,235 @@
+
+/-!
+This updates the board in the `GameState` with the new tile, and then changes the current player,
+providing no output (`Unit` return type).
+
+So finally, you can combine these functions together with `do` notation, and it actually looks quite
+clean! You don't need to worry about the side effects. The different monadic functions handle them.
+Here's a sample of what your function might look like to play one turn of the game. At the end, it
+returns a boolean determining if all the spaces have been filled.
+-/
+
+def isGameDone : StateM GameState Bool := do
+  return (← findOpen).isEmpty
+
+def nextTurn : StateM GameState Bool := do
+  let i ← chooseRandomMove
+  applyMove i
+  isGameDone
+
+/-!
+To give you a quick test harness that runs all moves for both players you can run this:
+-/
+
+def initBoard : Board := Id.run do
+  let mut board := HashMap.empty
+  for i in [0:3] do
+    for j in [0:3] do
+      let t : TileIndex := (i, j)
+      board := board.insert t TileEmpty
+  board
+
+def printBoard (board : Board) : IO Unit := do
+  let mut row : List String := []
+  for i in board.toList do
+    let s := match i.2 with
+      | TileEmpty => " "
+      | TileX => "X"
+      | TileO => "O"
+    row := row.append [s]
+    if row.length == 3 then
+      IO.println row
+      row := []
+
+def main : IO Unit := do
+  let gen ← IO.stdGenRef.get
+  let (x, gen') := randNat gen 0 1
+  let mut gs : GameState := {
+    board := initBoard,
+    currentPlayer := if x = 0 then XPlayer else OPlayer,
+    generator := gen' }
+  for _ in [0:9] do
+    let (_, g) := nextTurn gs
+    gs := g
+  printBoard gs.board
+
+#eval main
+-- [X, X, O]
+-- [X, O, O]
+-- [O, O, X]
+
+/-!
+
+Note that when you run the above code interactively the random number generator always starts in the
+same place.  But if you run `lean --run states.lean` then you will see randomness in the result.
+
+## Implementation
+
+It may be helpful to see how the `StateM` monad adds the input state and output state.  If you look
+at the reduced Type for `nextTurn`:
+-/
+#reduce StateM GameState Bool
+-- GameState → Bool × GameState
+/-!
+
+So a function like `nextTurn` that might have just returned a `Bool` has been modified by the
+`StateM` monad such that the initial `GameState` is passed in as a new input argument, and the output
+value has been changed to the pair `Bool × GameState` so that it can return the pure `Bool` and the
+updated `GameState`.  This is why the call to `nextTurn` looks like this: `let (_, g) := nextTurn gs`.
+This expression `(_, g)` conveniently breaks the pair up into 2 values, it doesn't care what the first
+value is (hence the underscore `_`), but it does need the updated state `g` which you can then assign
+back to the mutable `gs` variable to use next time around this loop.
+
+## StateM vs ReaderM
+
+While `ReaderM` functions can use `withReader` to modify the state before calling another function,
```

Comment:
I have never seen `withReader f act` as modifying the `state`. For me, it is modifying the **context** for the monadic action `act`. Remark, some people use environment instead of context.

Reply:
Yes, great example of how using the term "context" instead of state here helps disambiguate.  I changed it.

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/8dd24bc9dae40eb19b4625bf0b94cfb3b15a2f07